### PR TITLE
Initial commit to get BERT + run_glue.py on TPU

### PR DESCRIPTION
**Pull request by jysohn23**
Verified performance numbers look at least comparable on a chip-to-chip basis (TPUv3 vs V100) for MRPC dataset (pretty much the same accuracy & f1 test metrics too). Runner script works for both GPU and TPU.

### CODE REVIEW

File: examples/run_glue.py [hunk 1]
```diff
@@ -548,5 +479,104 @@ def main():
...
+                        help="For fp16: Apex AMP optimization level selected in ['O0', 'O1', 'O2', and 'O3']."
+                             "See details at https://nvidia.github.io/apex/amp.html")
+    parser.add_argument("--local_rank", type=int, default=-1,
+                        help="For distributed training: local_rank")
+    parser.add_argument('--server_ip', type=str, default='', help="For distant debugging.")
+    parser.add_argument('--server_port', type=str, default='', help="For distant debugging.")
+
+    parser.add_argument('--use_tpu', action='store_true', help='Whether to use TPUs.')
+    parser.add_argument('--num_cores', default=8, type=int, help='Number of TPU cores to use.')
+    parser.add_argument('--metrics_debug', action='store_true', help='Whether to print debug metrics.')
```
Comment (taylanbil) [inline examples/run_glue.py:564]: are these the only tpu specific args?

File: examples/run_glue.py [hunk 2]
```diff
@@ -548,5 +479,104 @@ def main():
...
+    parser.add_argument("--no_cuda", action='store_true',
+                        help="Avoid using CUDA when available")
+    parser.add_argument('--overwrite_output_dir', action='store_true',
+                        help="Overwrite the content of the output directory")
+    parser.add_argument('--overwrite_cache', action='store_true',
+                        help="Overwrite the cached training and evaluation sets")
+    parser.add_argument('--seed', type=int, default=42,
+                        help="random seed for initialization")
+
+    parser.add_argument('--fp16', action='store_true',
```
Comment (taylanbil) [inline examples/run_glue.py:552]: Does enabling this break tpu? It did for fairseq.

File: examples/run_glue.py [hunk 3]
```diff
@@ -220,19 +247,23 @@ def evaluate(args, model, tokenizer, prefix=""):
...
         logger.info("***** Running evaluation {} *****".format(prefix))
-        logger.info("  Num examples = %d", len(eval_dataset))
+        logger.info("  Num examples = %d", num_batches * args.eval_batch_size)
         logger.info("  Batch size = %d", args.eval_batch_size)
         eval_loss = 0.0
         nb_eval_steps = 0
         preds = None
         out_label_ids = None
-        for batch in tqdm(eval_dataloader, desc="Evaluating"):
+        for batch in tqdm(eval_dataloader, desc="Evaluating", disable=args.use_tpu):
```
Comment (taylanbil) [inline examples/run_glue.py:266]: why disable this?

File: examples/run_glue.py [hunk 4]
```diff
@@ -505,7 +436,7 @@ def main():
 
 
     # Saving best-practices: if you use defaults names for the model, you can reload it using from_pretrained()
-    if args.do_train and (args.local_rank == -1 or torch.distributed.get_rank() == 0) and not args.tpu:
+    if args.do_train and (args.local_rank == -1 or torch.distributed.get_rank() == 0):
```
Comment (taylanbil) [inline examples/run_glue.py:439]: where do these get set correctly for our MP purposes?

File: examples/run_glue.py [hunk 5]
```diff
@@ -548,5 +479,104 @@ def main():
...
+
+def _tpu_mp_fn(rank, args):
+    global logger
+    logger_name ='{} [{}]'.format(__name__, xm.get_ordinal(defval=-1))
+    logger = logging.getLogger(logger_name)
+    main(args)
+
+def main_cli():
+    args = get_args()
+    if args.use_tpu:
```
Comment (taylanbil) [inline examples/run_glue.py:576]: having to pass `--use_tpu` every time feels annoying, but no big deal.

File: examples/run_glue.py [hunk 6]
```diff
@@ -176,35 +207,31 @@ def train(args, train_dataset, model, tokenizer):
...
+            xm.master_print(met.metrics_report())
         if args.max_steps > 0 and global_step > args.max_steps:
             train_iterator.close()
             break
 
     if args.local_rank in [-1, 0]:
         tb_writer.close()
 
-    return global_step, tr_loss / global_step
+    return global_step, loss.item()
```
Comment (taylanbil) [inline examples/run_glue.py:234]: is this equivalent?

File: examples/run_glue.py [hunk 7]
```diff
@@ -157,14 +185,17 @@ def train(args, train_dataset, model, tokenizer):
...
-            if (step + 1) % args.gradient_accumulation_steps == 0 and not args.tpu:
+            if (step + 1) % args.gradient_accumulation_steps == 0:
                 if args.fp16:
                     torch.nn.utils.clip_grad_norm_(amp.master_params(optimizer), args.max_grad_norm)
                 else:
                     torch.nn.utils.clip_grad_norm_(model.parameters(), args.max_grad_norm)
 
-                optimizer.step()
+                if args.use_tpu:
+                    xm.optimizer_step(optimizer, barrier=True)
```
Comment (taylanbil) [inline examples/run_glue.py:195]: why do we need barrier here? Isn't it in parallelloader already?

**Comment by jysohn23** [comment]
Thanks for the review @taylanbil!

Based on offline conversation with Google and HuggingFace teams, will close the PR in favor of preparing a separate `run_glue_tpu.py` runner.

File: examples/run_glue.py (hunk 7 above)
Comment (jysohn23) [reply examples/run_glue.py:195]: Good point! We don't need barrier here. Artifact I forgot to cleanup from testing on single core.

File: examples/run_glue.py (hunk 6 above)
Comment (jysohn23) [reply examples/run_glue.py:234]: Sort of, just that it's not a real average.

File: examples/run_glue.py (hunk 5 above)
Comment (jysohn23) [reply examples/run_glue.py:576]: Yep, it indeed is so will create separate runner as discussed offline.
